pkg/reconciler/serverlessservice/resources/services.go

```diff
@@ -57,14 +57,46 @@ func MakePublicService(sks *v1alpha1.ServerlessService) *corev1.Service {
 				Protocol:   corev1.ProtocolTCP,
 				Port:       int32(networking.ServicePort(sks.Spec.ProtocolType)),
 				TargetPort: targetPort(sks),
+			}, {
+				Name:       networking.ServicePortName(sks.Spec.ProtocolType) + "-proxy",
+				Protocol:   corev1.ProtocolTCP,
+				Port:       9999,
+				TargetPort: targetPort(sks),
 			}},
 		},
 	}
 }
 
 // MakePublicEndpoints constructs a K8s Endpoints that is not backed a selector
 // and will be manually reconciled by the SKS controller.
-func MakePublicEndpoints(sks *v1alpha1.ServerlessService, src *corev1.Endpoints) *corev1.Endpoints {
+func MakePublicEndpoints(sks *v1alpha1.ServerlessService, activatorEps *corev1.Endpoints, appEps *corev1.Endpoints) *corev1.Endpoints {
+	servePort := corev1.EndpointPort{
+		Name: networking.ServicePortName(sks.Spec.ProtocolType),
+		Port: targetPort(sks).IntVal,
+	}
+	proxyPort := corev1.EndpointPort{
+		Name: networking.ServicePortName(sks.Spec.ProtocolType) + "-proxy",
+		Port: targetPort(sks).IntVal,
+	}
+
+	activatorPorts := []corev1.EndpointPort{servePort}
+	appPorts := []corev1.EndpointPort{proxyPort}
+	if sks.Spec.Mode == v1alpha1.SKSOperationModeServe {
```

Comment:
you lost the logic here, where we keep activator pods if the mode is serve, but no ready app endpoints are ready still (scale from 0).